flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp

```diff
@@ -426,6 +426,93 @@ static bool hasPointerType(mlir::Type type) {
   return type.isa<fir::PointerType>();
 }
 
+// This is a NF performance hack. It makes a simple test that the slices of the
+// load, \p ld, and the merge store, \p st, are trivially mutually exclusive.
+static bool mutuallyExclusiveSliceRange(ArrayLoadOp ld, ArrayMergeStoreOp st) {
+  // If the same array_load, then no further testing is warranted.
+  if (ld.getResult() == st.original())
+    return false;
+
+  auto getSliceOp = [](mlir::Value val) -> fir::SliceOp {
+    if (!val)
+      return {};
+    auto sliceOp = mlir::dyn_cast_or_null<fir::SliceOp>(val.getDefiningOp());
+    if (!sliceOp)
+      return {};
+    return sliceOp;
+  };
+
+  auto ldSlice = getSliceOp(ld.slice());
+  auto stSlice = getSliceOp(st.slice());
+  if (!ldSlice || !stSlice)
+    return false;
+
+  // Resign on subobject slices.
```

Comment:
If the goal is to return a 'definitely no overlap' result, wouldn't `field` or `substr` analysis help, but never hurt?  That is, any dimension of the access that doesn't overlap implies an overall non-overlap result, no?

Reply:
Substrings are not obvious, since they can be abused with out of range values. It didn't seem worth the effort in this hacky solution.